src/mapblock.cpp

```diff
@@ -222,16 +223,64 @@ void MapBlock::copyFrom(const VoxelManipulator &src)
 	v3s16 data_size(MAP_BLOCKSIZE, MAP_BLOCKSIZE, MAP_BLOCKSIZE);
 	VoxelArea data_area(v3s16(0,0,0), data_size - v3s16(1,1,1));
 
+	expandNodesIfNeeded();
 	// Copy from VoxelManipulator to data
 	src.copyTo(data, data_area, v3s16(0,0,0),
 			getPosRelative(), data_size);
+	tryShrinkNodes();
+}
+
+void MapBlock::reallocate(u32 count, MapNode n)
```

Comment:
Another option is to pass a boolean here (mono block on or off). Not sure it's cleaner, but since there are hardcoded checks count == 1 in there it might.

Reply:
I would add an `assert(count == 1 || count == nodecount)`